engine/consensus/sealing/core.go

```diff
@@ -36,999 +28,394 @@ import (
-	for _, sealedID := range sealedIDs {
-		clear[sealedID] = true
-	}
-
-	sealed, err := c.state.Sealed().Head()
+	checkEmergencySealingSpan := c.tracer.StartSpanFromParent(processFinalizedBlockSpan, trace.CONSealingCheckForEmergencySealableBlocks)
+	// check if there are stale results qualified for emergency sealing
+	err = c.checkEmergencySealing(lastSealed.Height, finalized.Height)
+	checkEmergencySealingSpan.Finish()
 	if err != nil {
-		return fmt.Errorf("could not get sealed head: %w", err)
+		return fmt.Errorf("could not check emergency sealing at block %v", finalizedBlockID)
 	}
 
-	// build a helper function that determines if an entity should be cleared
-	// if it references the block with the given ID
-	missingIDs := make(map[flow.Identifier]bool) // count each missing block only once
-	shouldClear := func(blockID flow.Identifier) (bool, error) {
-		if c.missing[blockID] >= 1000 {
-			return true, nil // clear if block is missing for 1000 seals already
-		}
-		header, err := c.headersDB.ByBlockID(blockID)
-		if errors.Is(err, storage.ErrNotFound) {
-			missingIDs[blockID] = true
-			return false, nil // keep if the block is missing, but count times missing
-		}
-		if err != nil {
-			return false, fmt.Errorf("could not check block expiry: %w", err)
-		}
-		if header.Height <= sealed.Height {
-			return true, nil // clear if sealed block is same or higher than referenced block
-		}
-		return false, nil
-	}
-
-	// The receipts mempool is aware of the Execution Tree structure formed by the execution results.
-	// It supports pruning by height: only results descending from the latest sealed and finalized
-	// result are relevant. Hence, we can prune all results for blocks _below_ the latest block with
-	// a finalized seal. Results of sufficient height for forks that conflict with the finalized fork
-	// are retained in the mempool. However, such orphaned forks do not grow anymore and their
-	// results will be progressively flushed out with increasing sealed-finalized height.
-	err = c.receipts.PruneUpToHeight(sealed.Height)
+	updateCollectorTreeSpan := c.tracer.StartSpanFromParent(processFinalizedBlockSpan, trace.CONSealingUpdateAssignmentCollectorTree)
+	// finalize forks to stop collecting approvals for orphan collectors
+	err = c.collectorTree.FinalizeForkAtLevel(finalized, lastSealed)
 	if err != nil {
-		return fmt.Errorf("failed to clean receipts mempool: %w", err)
+		updateCollectorTreeSpan.Finish()
+		return fmt.Errorf("collectors tree could not finalize fork: %w", err)
 	}
 
-	// for each memory pool, clear if the related block is no longer relevant or
-	// if the seal was already built for it (except for seals themselves)
-	for _, result := range c.incorporatedResults.All() {
-		remove, err := shouldClear(result.Result.BlockID)
+	// pruning of collectors tree makes sense only with values that are increasing
+	// passing value lower than before is not supported by collectors tree
+	if lastSealedHeightIncreased {
```

Comment:
⚠️ 
I am worried that this might crash the node:
* assume two threads (`A` and `B`) are concurrently finalizing. Thread `A` finalizes block at height 10 and `B` the block at height 11. 
* Up to this point, `A` went first and subsequently `B`. Hence, they both have local values `lastSealedHeightIncreased = true`
* But for the `if` statement, thread `B` prunes first and then thread `A`. Under this condition, the collector tree will error 

#### Suggestion:
* We could change the implementation of  `AssignmentCollectorTree`, to not error if the respective height has already been pruned.
* then, we  also wouldn't need this if statement here anymore